lib/services/role.go

```diff
@@ -1095,6 +1095,50 @@ func (set RoleSet) PinSourceIP() bool {
 	return false
 }
 
+// MFAParams returns MFA params for the given user given their roles, the cluster
+// auth preference, and whether mfa has been verified.
+func (set RoleSet) MFAParams(authPrefRequirement types.RequireMFAType) (params AccessMFAParams) {
+	if authPrefRequirement == types.RequireMFAType_HARDWARE_KEY_TOUCH {
+		// per-session MFA is overridden by hardware key PIV touch requirement.
+		return AccessMFAParams{
+			NeverRequired: true,
+		}
+	}
+
+	if authPrefRequirement.IsSessionMFARequired() {
+		params.AlwaysRequired = true
+	} else {
+		params.NeverRequired = true
+	}
```

Comment:
This look like something that can be simplified. Please look at my other comment below.